File: 3.1 Document Preprocessing/go/main.go
// ============================================================================
//  DOCUMENT PREPROCESSING PIPELINE  (Go)
// ============================================================================
//
//  GOAL
//  ----
//  In a RAG (Retrieval-Augmented Generation) system, everything begins with
//  RAW TEXT. But your knowledge lives in many formats: PDFs, Word docs,
//  Markdown notes, plain text...
//
//  This program does ONE job: scan the ./docs folder and, for EVERY file it
//  finds, extract the plain text regardless of its format.
//
//  THE MENTAL MODEL
//  ----------------
//      docs/ folder -> discover files -> choose the right extractor
//                                        (based on file extension)
//                   -> extract text   -> collect results
//
//  Each FORMAT is stored very differently on disk:
//    - .txt / .md  -> already plain text, just read the bytes
//    - .pdf        -> binary; a library walks its internal structure
//    - .docx       -> secretly a ZIP archive of XML files; we unzip it
//                     ourselves using Go's standard library (educational!)
//  We hide those differences behind one idea: every extractor is a function
//  that takes a file path and returns a string. That uniformity is the trick.
// ============================================================================

package main

import (
	"archive/zip"       // to open .docx files, which are really ZIP archives
	"encoding/xml"      // to parse the XML inside a .docx
	"fmt"               // formatted printing
	"io"                // reading streams of bytes
	"os"                // file system access
	"path/filepath"     // OS-independent path handling
	"strings"           // string helpers (Join, ToLower, Fields...)

	"github.com/ledongthuc/pdf" // third-party: extracts text from PDF files
)

// ExtractedDocument is a small record describing one processed file.
// In Go we define a struct to give our data named, typed fields.
type ExtractedDocument struct {
	FileName  string // e.g. "report.pdf"
	FileType  string // the extension, e.g. ".pdf"
	Text      string // the extracted plain text
	CharCount int    // quick sanity check on how much we got
}

// docsDir is the folder we scan, relative to where you run the program.
const docsDir = "docs"

// ===========================================================================
//  THE EXTRACTORS — one function per format.
//  They all share the SAME signature: (path string) (string, error).
//  Returning an error is the idiomatic Go way to say "this might fail".
// ===========================================================================

// extractPlainText handles .md and .txt: they're ALREADY text on disk, so
// "extraction" is just reading the file's bytes and converting to a string.
func extractPlainText(path string) (string, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

// extractPDF handles .pdf (binary). The pdf library opens the file, then we
// loop over its pages and append each page's plain text.
func extractPDF(path string) (string, error) {
	// pdf.Open returns an *os.File (which we must close) and a Reader.
	f, reader, err := pdf.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close() // `defer` runs this when the function returns — guaranteed cleanup.

	var builder strings.Builder // efficient way to assemble a big string piece by piece.
	totalPages := reader.NumPage()
	for pageIndex := 1; pageIndex <= totalPages; pageIndex++ {
		page := reader.Page(pageIndex)
		if page.V.IsNull() {
			continue // skip empty/invalid pages
		}
		// GetPlainText pulls the readable text out of the page's content.
		content, err := page.GetPlainText(nil)
		if err != nil {
			return "", err
		}
		builder.WriteString(content)
	}
	return builder.String(), nil
}

// extractDOCX handles .docx. A .docx is a ZIP archive; the actual text lives
// in a file inside it called "word/document.xml". We:
//   1) open the ZIP,
//   2) find word/document.xml,
//   3) parse the XML and pull out every <w:t> ("text") element.
// Doing this by hand (instead of a library) shows you what a .docx REALLY is.
func extractDOCX(path string) (string, error) {
	// Open the .docx as a ZIP archive.
	archive, err := zip.OpenReader(path)
	if err != nil {
		return "", err
	}
	defer archive.Close()

	// Find the document body inside the archive.
	var docXML io.ReadCloser
	for _, file := range archive.File {
		if file.Name == "word/document.xml" {
			docXML, err = file.Open()
			if err != nil {
				return "", err
			}
			break
		}
	}
	if docXML == nil {
		return "", fmt.Errorf("word/document.xml not found inside %s", path)
	}
	defer docXML.Close()

	// We stream through the XML token by token. Every run of visible text in
	// Word is wrapped in a <w:t> element, so we collect the characters that
	// appear inside those elements.
	decoder := xml.NewDecoder(docXML)
	var builder strings.Builder
	insideTextElement := false

	for {
		token, err := decoder.Token()
		if err == io.EOF {
			break // reached the end of the XML
		}
		if err != nil {
			return "", err
		}

		switch element := token.(type) {
		case xml.StartElement:
			// "t" is the local name of Word's <w:t> text element.
			if element.Name.Local == "t" {
				insideTextElement = true
			}
		case xml.EndElement:
			if element.Name.Local == "t" {
				insideTextElement = false
			}
		case xml.CharData:
			// CharData is the raw text between tags. Keep it only when we're
			// currently inside a <w:t> element.
			if insideTextElement {
				builder.Write(element)
			}
		}
	}
	return builder.String(), nil
}

// ===========================================================================
//  THE DISPATCHER — "given a file, which extractor do I use?"
//  We map each extension to its extractor function. In Go, functions are
//  values, so we can store them in a map. This is a "dispatch table":
//  adding a new format later = add ONE entry here.
// ===========================================================================
var extractors = map[string]func(string) (string, error){
	".pdf":  extractPDF,
	".docx": extractDOCX,
	".md":   extractPlainText,
	".txt":  extractPlainText,
}

// extractText looks at the extension and runs the matching extractor.
func extractText(path string) (string, error) {
	// filepath.Ext("a/b/report.PDF") -> ".PDF"; ToLower normalizes the case
	// so ".PDF"/".Pdf"/".pdf" all route to the same extractor.
	ext := strings.ToLower(filepath.Ext(path))

	extractor, ok := extractors[ext]
	if !ok {
		// Return an error loudly instead of silently producing "" — you want
		// to KNOW when a file was skipped for being an unsupported type.
		return "", fmt.Errorf("unsupported file type: %q", ext)
	}
	return extractor(path)
}

// ===========================================================================
//  THE PIPELINE — tie it all together.
// ===========================================================================
func processAllDocuments() ([]ExtractedDocument, error) {
	// 1) DISCOVER: list everything inside the docs folder.
	entries, err := os.ReadDir(docsDir)
	if err != nil {
		return nil, err
	}

	var results []ExtractedDocument

	// 2) LOOP over each entry.
	for _, entry := range entries {
		name := entry.Name()

		// Skip sub-directories and hidden files (like macOS's .DS_Store).
		if entry.IsDir() || strings.HasPrefix(name, ".") {
			continue
		}

		fullPath := filepath.Join(docsDir, name)

		// 3) EXTRACT. If ONE file fails we log it and keep going, so a single
		//    bad file never crashes the whole run. Real folders are messy;
		//    a production pipeline degrades gracefully.
		text, err := extractText(fullPath)
		if err != nil {
			fmt.Printf("⚠️  Skipped %s: %v\n", name, err)
			continue
		}

		results = append(results, ExtractedDocument{
			FileName:  name,
			FileType:  strings.ToLower(filepath.Ext(name)),
			Text:      text,
			CharCount: len(text),
		})
		fmt.Printf("✅ %s — extracted %d characters\n", name, len(text))
	}

	return results, nil
}

func main() {
	fmt.Printf("📂 Scanning: %s\n\n", docsDir)

	docs, err := processAllDocuments()
	if err != nil {
		fmt.Printf("Pipeline failed: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("\n📊 Done. Processed %d document(s).\n", len(docs))

	// Print a short preview so you can SEE the text really came out.
	for _, doc := range docs {
		// strings.Fields splits on any whitespace; Join with single spaces
		// collapses newlines/tabs into a clean one-line preview.
		oneLine := strings.Join(strings.Fields(doc.Text), " ")
		preview := oneLine
		if len(preview) > 120 {
			preview = preview[:120] + " ..."
		}
		fmt.Printf("\n--- %s (%d chars) ---\n%s\n", doc.FileName, doc.CharCount, preview)
	}

	// NEXT STEPS in a full RAG system (not done here):
	//   text -> CHUNK into small pieces -> EMBED each chunk -> STORE in a
	//   vector database. This program stops at clean text: the foundation.
}

File: 3.1 Document Preprocessing/python/main.py
"""
============================================================================
 DOCUMENT PREPROCESSING PIPELINE  (Python)
============================================================================

 GOAL
 ----
 In a RAG (Retrieval-Augmented Generation) system, everything starts with
 RAW TEXT. But your knowledge lives in many formats: PDFs, Word docs,
 Markdown notes, plain text files...

 This script does ONE job: scan the ./docs folder and, for EVERY file it
 finds, pull out the plain text — regardless of the file's format.

 THE MENTAL MODEL
 ----------------
 It's an assembly line:

     docs/ folder ─▶ discover files ─▶ choose the right extractor
                                        (based on file extension)
                  ─▶ extract text   ─▶ collect results

 Each FORMAT needs its own extractor because the formats are stored very
 differently on disk:
   - .txt / .md  -> already plain text, just read it
   - .pdf        -> binary; a library walks its internal structure
   - .docx       -> secretly a ZIP of XML files; a library unpacks it
 We hide those differences behind a common idea: every extractor is a
 function that takes a file path and returns a string. That uniformity is
 the entire trick.
============================================================================
"""

# `pathlib` is the modern, object-oriented way to handle file paths in
# Python. A `Path` knows things about itself, like its `.suffix` (extension).
from pathlib import Path

# `dataclass` lets us define a small structured record with almost no
# boilerplate — just a labeled container for our results.
from dataclasses import dataclass

# Third-party libraries (installed via `pip install -r requirements.txt`).
# Each is a specialist that decodes one tricky format:
#   pypdf        -> reads PDF files
#   python-docx  -> reads .docx (Word) files (imported as `docx`)
from pypdf import PdfReader
from docx import Document


# ---------------------------------------------------------------------------
# A simple record describing one processed document.
# `@dataclass` auto-generates the __init__ etc. for us.
# ---------------------------------------------------------------------------
@dataclass
class ExtractedDocument:
    file_name: str   # e.g. "report.pdf"
    file_type: str   # the extension, e.g. ".pdf"
    text: str        # the extracted plain text
    char_count: int  # quick sanity check on how much we got


# The folder we scan. `__file__` is THIS script's path; `.parent` is the
# folder it lives in; then we point at the sibling "docs" folder.
DOCS_DIR = Path(__file__).parent / "docs"


# ===========================================================================
#  THE EXTRACTORS — one function per format.
#  They all share the SAME shape: take a Path, return a str. That sameness
#  is what lets the dispatcher treat them interchangeably.
# ===========================================================================

def extract_pdf(file_path: Path) -> str:
    """PDF — binary format. PdfReader parses its internal structure and
    lets us pull text page by page. We join the pages with newlines."""
    reader = PdfReader(str(file_path))
    # A PDF is a list of pages; extract_text() returns the text of each.
    # `or ""` guards against pages that have no extractable text (e.g.
    # scanned images), so we never try to join a None into the string.
    pages = [page.extract_text() or "" for page in reader.pages]
    return "\n".join(pages)


def extract_docx(file_path: Path) -> str:
    """DOCX — a Word document is actually a ZIP of XML files. python-docx
    unzips it and exposes the content as a list of paragraphs. We pull the
    text out of each paragraph and join them with newlines."""
    document = Document(str(file_path))
    paragraphs = [p.text for p in document.paragraphs]
    return "\n".join(paragraphs)


def extract_plain_text(file_path: Path) -> str:
    """Markdown (.md) and plain text (.txt) are ALREADY text on disk, so
    'extraction' is simply reading the file as UTF-8. We keep Markdown
    symbols (#, *, -) as-is; they're harmless for RAG and can even help
    preserve document structure."""
    # encoding="utf-8" decodes the bytes into a proper string.
    return file_path.read_text(encoding="utf-8")


# ===========================================================================
#  THE DISPATCHER — "given a file, which extractor do I use?"
#
#  We map each file extension to its extractor function. This is called a
#  "dispatch table". Adding a new format later = add ONE line to this dict.
#  Notice we store the FUNCTIONS themselves (no parentheses) as values.
# ===========================================================================
EXTRACTORS = {
    ".pdf": extract_pdf,
    ".docx": extract_docx,
    ".md": extract_plain_text,
    ".txt": extract_plain_text,
}


def extract_text(file_path: Path) -> str:
    """Look at the extension, find the matching extractor, and run it."""
    # `.suffix` gives ".PDF"; `.lower()` normalizes so ".PDF"/".Pdf"/".pdf"
    # all route to the same place.
    ext = file_path.suffix.lower()

    extractor = EXTRACTORS.get(ext)
    if extractor is None:
        # Raise loudly instead of silently returning "" — you want to KNOW
        # when a file was skipped because of an unsupported type.
        raise ValueError(f'Unsupported file type: "{ext}"')

    return extractor(file_path)


# ===========================================================================
#  THE PIPELINE — tie it all together.
# ===========================================================================
def process_all_documents() -> list[ExtractedDocument]:
    results: list[ExtractedDocument] = []

    # 1) DISCOVER + LOOP: iterdir() yields every entry in the folder.
    #    sorted() just makes the output order predictable.
    for file_path in sorted(DOCS_DIR.iterdir()):
        # Skip sub-folders and hidden files (like macOS's .DS_Store).
        if file_path.is_dir() or file_path.name.startswith("."):
            continue

        # 2) EXTRACT — wrapped in try/except so ONE bad file doesn't crash
        #    the whole run. Real folders are full of surprises; a production
        #    pipeline must degrade gracefully instead of exploding.
        try:
            text = extract_text(file_path)
            results.append(
                ExtractedDocument(
                    file_name=file_path.name,
                    file_type=file_path.suffix.lower(),
                    text=text,
                    char_count=len(text),
                )
            )
            print(f"✅ {file_path.name} — extracted {len(text)} characters")
        except Exception as err:
            # Report and move on to the next file.
            print(f"⚠️  Skipped {file_path.name}: {err}")

    return results


def main() -> None:
    print(f"📂 Scanning: {DOCS_DIR}\n")
    docs = process_all_documents()

    print(f"\n📊 Done. Processed {len(docs)} document(s).")

    # Show a short preview so you can SEE the text really came out.
    for doc in docs:
        # Collapse whitespace and trim to 120 chars for a clean one-liner.
        preview = " ".join(doc.text.split())[:120]
        print(f"\n--- {doc.file_name} ({doc.char_count} chars) ---")
        print(preview + (" ..." if doc.char_count > 120 else ""))

    # NEXT STEPS in a full RAG system (not done here):
    #   text -> CHUNK into small pieces -> EMBED each chunk -> STORE in a
    #   vector database. This script stops at clean text: the foundation.


# This guard means "only run main() when the file is executed directly"
# (i.e. `python main.py`), not when it's imported by another module.
if __name__ == "__main__":
    main()
